### PR TITLE
Use "dylink" section for main module dependencies

### PR DESCRIPTION
**Pull request by sbc100**
We already do this for side modules so it seems consistent
and cleaner.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -2860,8 +2860,7 @@ def do_binaryen(target, options, wasm_target):
     building.eval_ctors(final_js, wasm_target, debug_info=intermediate_debug_info)
 
   # after generating the wasm, do some final operations
-  if shared.Settings.SIDE_MODULE:
-    webassembly.add_dylink_section(wasm_target, shared.Settings.RUNTIME_LINKED_LIBS)
+  webassembly.add_dylink_section(wasm_target, shared.Settings.RUNTIME_LINKED_LIBS)
```
Comment (kripken) [inline emcc.py:2863]: shouldn't this check for main or side modules, as we don't want it otherwise?

Comment (sbc100) [reply emcc.py:2863]: Oh course, doh!

File: emcc.py [hunk 2]
```diff
@@ -2860,8 +2860,8 @@ def do_binaryen(target, options, wasm_target):
     building.eval_ctors(final_js, wasm_target, debug_info=intermediate_debug_info)
 
   # after generating the wasm, do some final operations
-  if shared.Settings.SIDE_MODULE:
-    webassembly.add_dylink_section(wasm_target, shared.Settings.RUNTIME_LINKED_LIBS)
+  if shared.Settings.RELOCATABLE:
```
Comment (dschuff) [inline emcc.py:2863]: (sorry for the "basic" questions here)... so RELOCATABLE applies to main modules and side modules?

File: src/preamble.js [hunk 3]
```diff
@@ -987,10 +989,11 @@ function createWasm() {
     exportAsmFunctions(exports);
 #endif
 #if USE_PTHREADS
-    // We now have the Wasm module loaded up, keep a reference to the compiled module so we can post it to the workers.
-    wasmModule = module;
```
Comment (dschuff) [inline src/preamble.js:991]: is this change just a drive-by?

**Review by dschuff — COMMENTED** [review]
was there just no library deps info on main modules before?

File: emcc.py (hunk 2 above)
Comment (sbc100) [reply emcc.py:2863]: Yes that is correct.

I'm actually planning on renaming `RELOCATABLE` to `DYNAMIC` to give it a better name going forward.

**Comment by sbc100** [comment]
> was there just no library deps info on main modules before?

Yes, but this was made up for by directly embedding the deps in the JS file instead.

File: src/preamble.js (hunk 3 above)
Comment (sbc100) [reply src/preamble.js:991]: I actually needed to revert this part..

File: emcc.py (hunk 2 above)
Comment (dschuff) [reply emcc.py:2868]: Thanks, yeah that's a much better name (even if it's just an internal define)

**Comment by dschuff** [comment]
Thanks. I agree that embedding the deps for dynamic executables similarly to dynamic libs is better than baking it into the JS. I guess the other parts of the dylink section will be different (e.g. dynamic executables won't need memory and tablesize). I guess the symbol table mechanism would be optional for executables (since we nee something like export-dynamic when libc is embedded?)

**Comment by sbc100** [comment]
> Thanks. I agree that embedding the deps for dynamic executables similarly to dynamic libs is better than baking it into the JS. I guess the other parts of the dylink section will be different (e.g. dynamic executables won't need memory and tablesize). I guess the symbol table mechanism would be optional for executables (since we need something like export-dynamic when libc is embedded?)

Actually main modules already have those other parts of the dylink section.. we were just not reading it in case of main module.

And yes, this is a first step towards extending the dylink section... actually reading it when loading the main module as well as the side modules.

File: src/worker.js [hunk 4]
```diff
@@ -45,9 +45,9 @@ Module['instantiateWasm'] = function(info, receiveInstance) {
   // Instantiate from the module posted from the main thread.
   // We can just use sync instantiation in the worker.
   var instance = new WebAssembly.Instance(Module['wasmModule'], info);
+  receiveInstance(instance, Module['wasmModule']);
```
Comment (kripken) [inline src/worker.js:48]: How about ifdefing here to avoid always passing `Module['wasmModule']`? we only need it when RELOCATABLE, and this is something I doubt closure will eliminate (but maybe worth checking)

Comment (sbc100) [reply src/worker.js:48]: OK.. seems a little excessive to me and I guess should also update the other callsites too though for consistency?

Comment (sbc100) [reply src/worker.js:49]: Done

**Review by kripken — APPROVED** [review]
Thanks!

Ah, now I remember that closure issue... yeah, annoying, but until it's fixed I think it's worth working around.
